notebooks/ch16/classification_sgd.md

```diff
@@ -0,0 +1,83 @@
+\nabla_\theta L(\theta, y) \approx -(y_i - \sigma_i)X_i
+$$
+
+We can then plug this back into the general formula for gradient descent to find estimates for $\theta$:
+
+$$
+\begin{align}
+\theta_{t+1} &= \theta_t - \alpha \cdot \nabla_\theta L(\theta, y) \\
+&\approx \theta_t + \alpha \cdot (y_i - \sigma_i)X_i
+\end{align}
+$$
+
+## Visualizing Stochastic Gradient Descent
+
+Below are visual examples of loss minimization in batch gradient descent and stochastic gradient descent.
+
+![](gd.png)
+![](sgd.png)
+
+At each iteration of batch gradient descent, we move in the direction of the true gradient of the loss (shown by the ellipses). On the other hand, at each iteration of stochastic gradient, our steps may not lead us in the direction of the true gradient; however, our $\theta$ estimates eventually converge to the true loss-minimizing values. At large values of $n$, this is much more desirable due to the fast speed and low computational cost at which we can generate estimates at each iteration.
+
+# Mini-batch Gradient Descent
+
+We can improve our estimates by increasing the number of observations that we select at each iteration. In **mini-batch gradient descent**, we take a simple random sample (known as a mini-batch) of observations and use the average of their loss gradients to construct an estimate of the true gradient of the cross entropy cost. Since our sample is randomly selected, the expectation of this estimate is equal to the true gradient. This is illustrated in the approximation below, where $\mathcal{B}$ is the mini-batch of data points that we randomly sample from the $n$ observations.
+
+$$
+\nabla_\theta L(\theta, y) \approx -\frac{1}{|\mathcal{B}|} \sum_{i\in\mathcal{B}}(y_i - \sigma_i)X_i
+$$
+
+In order to select our mini-batch, we could randomly sample at each iteration. In practice, we randomly shuffle the training data and select our mini-batch by looping through the shuffled data. This also applies to stochastic gradient descent.
+
+
+## Selecting the Mini-Batch Size
+
+The mini-batch size is often set to small numbers such as 1 or 2 to minimize computational cost. However, when working with parallel machines such as a GPU or a distributed computing system, we can increase the mini-batch size to take advantage of the hardware. Since mini-batches can be computed in parallel, the runtime of larger mini-batches is roughly equivalent to the runtime of stochastic gradient descent.
+
+# Summary
+
+Batch gradient descent allows us to minimize the loss of the cost function at each iteration and take direct steps towards the loss-minimizing $\theta$ parameters. However, using the entire batch of $n$ observations can become computationally expensive for large $n$; in such scenarios, we use stochastic gradient descent or mini-batch gradient descent. While they may take noisier paths to the loss-minimizing $\theta$ parameters, they also reduce runtime by only computing the loss gradient over a small mini-batch of observations.
```

Comment:
nit: GD doesn't minimize the loss "at each iteration"; it moves \theta in the direction that lowers cost. "in such scenarios" is bit wordy. In the last sentence, "they" is a bit ambiguous too. We don't really say that paths are "noisier" than others. How about:

We use gradient descent to iteratively improve model parameters until the model achieves low cost. Since batch gradient descent is computationally intractable with large datasets, we often use stochastic gradient descent to fit models instead. With a GPU, mini-batch gradient descent can converge more quickly than stochastic gradient descent.

Although stochastic gradient descent typically takes more individual updates to converge than batch gradient descent, stochastic gradient descent often converges more quickly because it computes updates faster.

Reply:
Fixed